src/main/java/org/broadinstitute/hellbender/transformers/IlluminaAdapterTrimTransformer.java

```diff
@@ -0,0 +1,56 @@
+package org.broadinstitute.hellbender.transformers;
+
+import htsjdk.samtools.util.SequenceUtil;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.util.List;
+
+/**
+ * Trims adapter sequences from read ends. If the adapter sequence ends before the end of the read, trims the remaining
+ * bases as well.
+ */
+public class IlluminaAdapterTrimTransformer implements ReadTransformer {
+    public static final long serialVersionUID = 1L;
+
+    private final int maxMismatches;
+    private final int minClipLength;
+    private final List<String> adapterSequences;
+
+    public IlluminaAdapterTrimTransformer(final int maxMismatches, final int minClipLength, final List<String> adapterSequences) {
+        this.maxMismatches = maxMismatches;
+        this.minClipLength = minClipLength;
+        this.adapterSequences = adapterSequences;
+    }
+
+    @Override
+    public GATKRead apply(final GATKRead read) {
+        if (read.getLength() < minClipLength) return read;
+        final byte[] bases = read.getBases();
+
+        //Adapter list loop
+        for (final String adapterSequence : adapterSequences) {
+            //Start at the end of the read and walk backwards
+            int alignmentLength = minClipLength;
+            for (int alignmentStart = bases.length - minClipLength; alignmentStart >= 0; alignmentStart--) {
+                int numMismatches = 0;
+                //Check each base for mismatches
+                for (int j = 0; j < alignmentLength; j++) {
+                    if (!SequenceUtil.isNoCall((byte)adapterSequence.charAt(j)) && bases[alignmentStart + j] != adapterSequence.charAt(j)) {
+                        if (++numMismatches > maxMismatches) break;
+                    }
+                }
+                if (numMismatches <= maxMismatches) {
```

Comment:
I think what you are saying is that if some adapter ends with the sequence of another adapter, or if a particular adapter ends with itself e.g.:

1) `AGCAT`
2) `GAGCAT`

Both end in `AGCAT` but if we had the read ending in `GAGCAT` and (2) happens to be tested before (1), it would only trim 4 bases instead of 5. Also if you had an adapter `AAAAA` and minimum clip length of 4, it would only clip 4 bases.

This is a very good point I didn't consider. I don't believe this is the case for any standard adapter sequences (Picard's MarkIlluminaAdapters also doesn't seem to account for it), but since this class exposes the use of custom adapters we should check for it.